File: app/model/bool/BoolNode.scala
package model.bool

import model.bool.BoolNode._

sealed abstract class BoolNode {

  def apply(assignment: BoolTableRow): Boolean

  def negate: BoolNode

  def getAsString(needsParans: Boolean): String

  def usedVariables: Set[Variable]

  def and(that: BoolNode): AndBoolNode = AndBoolNode(this, that)

  def or(that: BoolNode): OrBoolNode = OrBoolNode(this, that)

  def nand(that: BoolNode): NAndBoolNode = NAndBoolNode(this, that)

  def nor(that: BoolNode): NOrBoolNode = NOrBoolNode(this, that)

  def xor(that: BoolNode): XOrBoolNode = XOrBoolNode(this, that)

  def impl(that: BoolNode): Implication = Implication(this, that)

  def equiv(that: BoolNode): Equivalency = Equivalency(this, that)

  def unary_-(): NotBoolNode = NotBoolNode(this)

  override def toString: String = getAsString(false)

  def asString: String = getAsString(false)

  def asHtml: String = {

    @annotation.tailrec
    def go(formula: String, replacers: List[(String, String)]): String = replacers match {
      case Nil                  => formula
      case (key, value) :: tail => go(formula.replaceAll(key, value), tail)
    }

    go(getAsString(false), HtmlReplacers.toList)
  }

}

object BoolNode {

  val HtmlReplacers: Map[String, String] = Map[String, String](
    "impl" -> "&rArr;",
    "nor" -> "&#x22bd;",
    "nand" -> "&#x22bc;",
    "equiv" -> "&hArr;",
    "not " -> "&not;",
    "and" -> "&and;",
    "xor" -> "&oplus;",
    "or" -> "&or;")

  def not(child: BoolNode): NotBoolNode = NotBoolNode(child)

  def constant(value: Boolean): Constant = if (value) TRUE else FALSE

}

// Not

final case class NotBoolNode(child: BoolNode) extends BoolNode {

  override def apply(assignment: BoolTableRow): Boolean = !child(assignment)

  override def negate: BoolNode = child

  override def getAsString(needsParans: Boolean): String = {
    val inner = "not " + child.getAsString(needsParans)
    if (needsParans) "(" + inner + ")" else inner
  }

  override def usedVariables: Set[Variable] = child.usedVariables

}

// Constants

sealed abstract class Constant(value: Boolean) extends BoolNode {

  override def apply(assignment: BoolTableRow): Boolean = value

  override def negate: BoolNode = if (value) FALSE else TRUE

  override def getAsString(needsParans: Boolean): String = if (value) "1" else "0"

  override def usedVariables: Set[Variable] = Set.empty

}

object FALSE extends Constant(false)

object TRUE extends Constant(true)

// Variable nodes

final case class Variable(variable: Char) extends BoolNode with Ordered[Variable] {

  override def apply(assignment: BoolTableRow): Boolean = assignment(this)

  override def negate: BoolNode = not(this)

  override def getAsString(needsParans: Boolean): String = variable.toString

  override def usedVariables: Set[Variable] = Set(this)

  override def compare(that: Variable): Int = variable - that.variable

}

// Binary nodes

sealed abstract class BinaryBoolNode(operator: String, left: BoolNode, right: BoolNode, eval: (Boolean, Boolean) => Boolean) extends BoolNode {

  override def apply(assignment: BoolTableRow): Boolean = eval.apply(left(assignment), right(assignment))

  override def getAsString(needsParans: Boolean): String = {
    def isBinaryNode(node: BoolNode): Boolean = node match {
      case _: BinaryBoolNode => true
      case _                 => false
    }

    val inner = left.getAsString(isBinaryNode(left)) + " " + operator.toLowerCase + " " + right.getAsString(isBinaryNode(right))
    if (needsParans) "(" + inner + ")" else inner
  }

  override def usedVariables: Set[Variable] = left.usedVariables ++ right.usedVariables

}


final case class OrBoolNode(l: BoolNode, r: BoolNode) extends BinaryBoolNode("OR", l, r, (l, r) => l || r) {

  override def negate: BoolNode = l nor r

}

final case class AndBoolNode(l: BoolNode, r: BoolNode) extends BinaryBoolNode("AND", l, r, (l, r) => l && r) {

  override def negate: BoolNode = l nand r

}

final case class XOrBoolNode(l: BoolNode, r: BoolNode) extends BinaryBoolNode("XOR", l, r, (l, r) => l ^ r) {

  override def negate: BoolNode = l equiv r

}


final case class NAndBoolNode(l: BoolNode, r: BoolNode) extends BinaryBoolNode("NAND", l, r, (l, r) => !(l && r)) {

  override def negate: BoolNode = l and r

}

final case class Implication(l: BoolNode, r: BoolNode) extends BinaryBoolNode("IMPL", l, r, (l, r) => !l || r) {

  override def negate: BoolNode = l and r.negate

}

final case class Equivalency(l: BoolNode, r: BoolNode) extends BinaryBoolNode("EQUIV", l, r, (l, r) => l == r) {

  override def negate: BoolNode = l xor r

}

final case class NOrBoolNode(l: BoolNode, r: BoolNode) extends BinaryBoolNode("NOR", l, r, (l, r) => !(l || r)) {

  override def negate: BoolNode = l or r

}

File: react-ui/src/randomTools/boolModel/boolNode.ts
import {Assignment} from './assignment';

export type BooleanNode = BooleanConstant | BooleanVariable | BooleanNot | BooleanBinaryNode;

export function getSubNodes(node: BooleanNode): BooleanNode[] {
  switch (node._type) {
    case 'Constant':
    case 'Variable':
      return [];
    case 'Not':
      return [node.child];
    default:
      return [node.left, node.right];
  }
}

export function evaluate(node: BooleanNode, assignments: Assignment): boolean {
  switch (node._type) {
    case 'Constant':
      return node.value;
    case 'Variable':
      return assignments[node.variable] || false;
    case 'Not':
      return !evaluate(node.child, assignments);
    default:
      return evaluateBinaryNode(node, assignments);
  }
}

function evaluateBinaryNode({_type, left, right}: BooleanBinaryNode, assignment: Assignment): boolean {
  const leftVal = evaluate(left, assignment);
  const rightVal = evaluate(right, assignment);

  return {
    'And': leftVal && rightVal,
    'Or': leftVal || rightVal,
    'NAnd': !(leftVal && rightVal),
    'NOr': !(leftVal || rightVal),
    'XOr': (leftVal && !rightVal) || (!leftVal && rightVal),
    'Equiv': leftVal === rightVal,
    'Impl': !leftVal || rightVal,
  }[_type];
}

export function stringifyNode(node: BooleanNode, needsParentheses = false): string {
  if (node._type === 'Constant') {
    return node.value ? '1' : '0';
  } else if (node._type === 'Variable') {
    return node.variable;
  } else if (node._type === 'Not') {
    return 'not ' + stringifyNode(node.child, true);
  } else {
    const child = stringifyNode(node.left, true) + ' ' + node._type.toLowerCase() + ' ' + stringifyNode(node.right, true);

    return needsParentheses
      ? '(' + child + ')'
      : child;
  }
}

export function getVariables(node: BooleanNode): BooleanVariable[] {
  // FIXME: test!
  switch (node._type) {
    case 'Constant':
      return [];
    case 'Variable':
      return [node];
    case 'Not':
      return getVariables(node.child);
    default:
      return Array.from(new Set([...getVariables(node.left), ...getVariables(node.right)]));
  }
}

export interface BooleanVariable {
  _type: 'Variable';
  variable: string;
}

export function booleanVariable(variable: string): BooleanVariable {
  return {_type: 'Variable', variable};
}


export interface BooleanConstant {
  _type: 'Constant';
  value: boolean;
}

export const BooleanTrue: BooleanConstant = {_type: 'Constant', value: true};

export const BooleanFalse: BooleanConstant = {_type: 'Constant', value: false};

export interface BooleanNot {
  _type: 'Not';
  child: BooleanNode;
}

export function not(child: BooleanNode): BooleanNot {
  return {_type: 'Not', child};
}

// Boolean binary nodes

type BooleanBinaryNodeType = 'And' | 'Or' | 'NAnd' | 'NOr' | 'XOr' | 'Equiv' | 'Impl';

export interface BooleanBinaryNode {
  _type: 'And' | 'Or' | 'NAnd' | 'NOr' | 'XOr' | 'Equiv' | 'Impl';
  left: BooleanNode;
  right: BooleanNode;
}

type BooleanBinaryNodeConstructor = (left: BooleanNode, right: BooleanNode) => BooleanBinaryNode;

function createBinaryNodeFunc(_type: BooleanBinaryNodeType): BooleanBinaryNodeConstructor {
  return (left, right) => ({_type, left, right});
}

export const and = createBinaryNodeFunc('And');
export const or = createBinaryNodeFunc('Or');
export const nand = createBinaryNodeFunc('NAnd');
export const nor = createBinaryNodeFunc('NOr');
export const xor = createBinaryNodeFunc('XOr');
export const equiv = createBinaryNodeFunc('Equiv');
export const impl = createBinaryNodeFunc('Impl');
